Type lowering and SILGen for imported ObjC async decls.

jckarter opened the pull request:
Type lowering and (todo) SILGen support for calling ObjC declarations imported as async.

jckarter committed:
SIL: Type lowering for imported ObjC async decls.

When lowering the type for `@objc` entry points of async declarations, restore
the original ObjC signature with the completion handler argument in the lowered
SIL type.

jckarter committed:
Add a mangling for completion block implementation functions.

DougGregor commented on lib/SIL/IR/SILFunctionType.cpp:
```diff
@@ -1738,17 +1773,34 @@ static bool isPseudogeneric(SILDeclRef c) {
+    return;
+  }
+  
+  // A foreign async convention means our lowered return type is Void, since
```
In the synchronous-error case I've been commenting on, you could end up with (e.g.) `BOOL` here, as dictated by the foreign error convention.

DougGregor reviewed (commented):
A few comments about some minor expectation mismatches with what the Clang importer is doing, but this is looking great!

jckarter committed:
SILGen: Stub out support for invoking foreign async methods.

Allow SILGen to not crash when invoking foreign async methods by emitting
`undef` for the completion callback going into the call, and for the
results that would be channeled back through awaiting the continuation.

jckarter commented on lib/SIL/IR/AbstractionPattern.cpp:
```diff
@@ -123,32 +124,40 @@ AbstractionPattern TypeConverter::getAbstractionPattern(EnumElementDecl *decl) {
+AbstractionPattern::EncodedForeignInfo::encode(
+                         const Optional<ForeignErrorConvention> &foreignError,
+                         const Optional<ForeignAsyncConvention> &foreignAsync) {
+  // Foreign async convention takes precedence.
+  if (foreignAsync.hasValue()) {
+    return EncodedForeignInfo(EncodedForeignInfo::Async,
+                              foreignAsync->completionHandlerParamIndex(),
+                              foreignAsync->completionHandlerErrorParamIndex());
+  } else if (foreignError.hasValue()) {
+    return EncodedForeignInfo(EncodedForeignInfo::Error,
+                              foreignError->getErrorParameterIndex(),
                               foreignError->isErrorParameterReplacedWithVoid(),
                               foreignError->stripsResultOptionality());
```
For now, I can have it throw up a diagnostic if this case comes up. It shouldn't be too hard to add support for handling either a synchronous or async error once the basic support is implemented.

DougGregor replied:
In my own experiments, this construct is very, very rare.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

DougGregor commented:
Those linux failures look completely unrelated:

```
<unknown>:0: error: compiled module was created by an older version of the compiler; rebuild 'Swift' and try again: /home/buildnode/jenkins/workspace/swift-PR-Linux@2/branch-main/buildbot_linux/swift-linux-x86_64/lib/swift_static/linux/Swift.swiftmodule/x86_64-unknown-linux-gnu.swiftmodule
```

Let's retry.

DougGregor commented:
@swift-ci please smoke test Linux
